pkg/agent/controller/bgp/controller.go

```diff
@@ -0,0 +1,950 @@
+	if err != nil {
+		return err
+	}
+	preRoutes := c.bgpPolicyState.routes
+	routesToAdvertise := curRoutes.Difference(preRoutes)
+	routesToWithdraw := preRoutes.Difference(curRoutes)
+
+	bgpServer := c.bgpPolicyState.bgpServer
+	for route := range routesToAdvertise {
+		if err := bgpServer.AdvertiseRoutes(c.ctx, []bgp.Route{route}); err != nil {
+			return err
+		}
+		c.bgpPolicyState.routes.Insert(route)
+	}
+	for route := range routesToWithdraw {
+		if err := bgpServer.WithdrawRoutes(c.ctx, []bgp.Route{route}); err != nil {
+			return err
+		}
+		c.bgpPolicyState.routes.Delete(route)
+	}
+
+	return nil
+}
+
+func hashNodeNameToIP(s string) string {
+	h := fnv.New32a() // Create a new FNV hash
+	h.Write([]byte(s))
+	hashValue := h.Sum32() // Get the 32-bit hash
+
+	// Convert the hash to a 4-byte slice
+	ip := make(net.IP, 4)
+	ip[0] = byte(hashValue >> 24)
+	ip[1] = byte(hashValue >> 16)
+	ip[2] = byte(hashValue >> 8)
+	ip[3] = byte(hashValue)
+
+	return ip.String()
+}
+
+func (c *Controller) getRouterID() (string, error) {
+	// According to RFC 4271:
+	// BGP Identifier:
+	//   This 4-octet unsigned integer indicates the BGP Identifier of
+	//   the sender.  A given BGP speaker sets the value of its BGP
+	//   Identifier to an IP address that is assigned to that BGP
+	//   speaker.  The value of the BGP Identifier is determined upon
+	//   startup and is the same for every local interface and BGP peer.
+	//
+	// In goBGP, only an IPv4 address can be used as the BGP Identifier (BGP router ID).
+	// For IPv4-only or dual-stack Kubernetes clusters, the Node's IPv4 address is used as the BGP router ID, ensuring
+	// uniqueness.
+	// For IPv6-only Kubernetes clusters without a Node IPv4 address, the router ID could be specified in the Node
+	// annotation `node.antrea.io/bgp-router-id`. If the annotation is not present, an IPv4 address will be generated by
+	// hashing the Node name and updated to the Node annotation `node.antrea.io/bgp-router-id`.
+
+	if c.enabledIPv4 {
+		return c.nodeIPv4Addr, nil
+	}
+
+	nodeObj, err := c.nodeLister.Get(c.nodeName)
+	if err != nil {
+		return "", fmt.Errorf("failed to get Node object: %w", err)
+	}
+
+	var exists bool
+	var routerID string
+	routerID, exists = nodeObj.GetAnnotations()[types.NodeBGPRouterIDAnnotationKey]
+	if !exists {
+		routerID = hashNodeNameToIP(c.nodeName)
+		patch, _ := json.Marshal(map[string]interface{}{
+			"metadata": map[string]interface{}{
+				"annotations": map[string]string{
+					types.NodeBGPRouterIDAnnotationKey: routerID,
+				},
+			},
+		})
+		if _, err := c.k8sClient.CoreV1().Nodes().Patch(context.TODO(), c.nodeName, apitypes.MergePatchType, patch, metav1.PatchOptions{}, "status"); err != nil {
+			return "", fmt.Errorf("failed to patch BGP router ID to Node annotation %s: %w", types.NodeBGPRouterIDAnnotationKey, err)
+		}
+	} else if !utilnet.IsIPv4String(routerID) {
+		return "", fmt.Errorf("BGP router ID should be an IPv4 address string")
```

Comment:
Should we also correct the routeID with the annotation NodeBGPRouterIDAnnotationKey in this case? Considering it's owned by Antrea but there might be a case that users modified it accidentally.

Reply:
I don't think we are obligated to correct that since that's what the users should do by themselves. cc @tnqn

Reply:
Logically we shouldn't update user input which would cause unexpected results. If users ever set this annotation, they want to specify the router ID to a value which is more meaningful for them, instead of using the system generated one. If they set a wrong value and we override it to a correct but different value, they wouldn't know they input a invalid ID and would think this is not configurable.

Reply:
I think it's fine as long as there is error log about it given this should only be used by advanced users who are familiar with BGP. And this is not the only case, the same situation applies to NPL, ServiceExternalIP, Service, Egress which takes configuration via user-provided annotation. Antrea users seem careful enough that we have never received bugs related to misconfigured annotations on other features :)